src/app/tests/TestICDManager.cpp

```diff
@@ -15,17 +15,211 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
+#include <app/EventManagement.h>
+#include <app/tests/AppTestContext.h>
+#include <lib/support/TestPersistentStorageDelegate.h>
+#include <lib/support/UnitTestContext.h>
 #include <lib/support/UnitTestRegistration.h>
 #include <nlunit-test.h>
+#include <system/SystemLayerImpl.h>
 
-int TestICDManager()
+#include <app/icd/ICDManager.h>
+#include <app/icd/ICDStateObserver.h>
+#include <app/icd/IcdManagementServer.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::System;
+
+namespace {
+
+class TestICDStateObserver : public app::ICDStateObserver
+{
+public:
+    void OnEnterActiveMode() {}
+};
+
+TestICDStateObserver mICDStateObserver;
+static Clock::Internal::MockClock gMockClock;
+static Clock::ClockBase * gRealClock;
+
+class TestContext : public Test::AppContext
+{
+public:
+    static int Initialize(void * context)
+    {
+        gRealClock = &SystemClock();
+        Clock::Internal::SetSystemClockForTesting(&gMockClock);
+
+        if (chip::DeviceLayer::PlatformMgr().InitChipStack() != CHIP_NO_ERROR)
+            return FAILURE;
+
+        if (AppContext::Initialize(context) != SUCCESS)
+            return FAILURE;
+
+        auto * ctx = static_cast<TestContext *>(context);
+
+        if (ctx->mEventCounter.Init(0) != CHIP_NO_ERROR)
+        {
+            return FAILURE;
+        }
+
+        ctx->mICDManager.Init(&ctx->testStorage, &ctx->GetFabricTable(), &mICDStateObserver);
+        return SUCCESS;
+    }
+
+    static int Finalize(void * context)
+    {
+        auto * ctx = static_cast<TestContext *>(context);
+        ctx->mICDManager.Shutdown();
+        app::EventManagement::DestroyEventManagement();
+        chip::System::Clock::Internal::SetSystemClockForTesting(gRealClock);
+
+        chip::DeviceLayer::PlatformMgr().Shutdown();
+        if (AppContext::Finalize(context) != SUCCESS)
+            return FAILURE;
+
+        return SUCCESS;
+    }
+
+    app::ICDManager mICDManager;
+
+private:
+    TestPersistentStorageDelegate testStorage;
+    MonotonicallyIncreasingCounter<EventNumber> mEventCounter;
+};
+
+} // namespace
+
+namespace chip {
+namespace app {
+class TestICDManager
 {
-    static nlTest sTests[] = { NL_TEST_SENTINEL() };
+public:
+    /*
+     * Advance the test Mock clock time by the amout passed in argument
+     * and then force the SystemLayer Timer event loop. It will check for any expired timer,
+     * and invoke their callbacks if there are any.
+     *
+     * @param time_ms: Value in milliseconds.
+     */
+    static void AdvanceClockAndRunEventLoop(uint32_t time_ms)
+    {
+        gMockClock.AdvanceMonotonic(System::Clock::Timeout(time_ms));
+#if CHIP_SYSTEM_CONFIG_USE_SOCKETS || CHIP_SYSTEM_CONFIG_USE_NETWORK_FRAMEWORK
```

Comment:
Why we need to drive the event loop by our-self?

Reply:
Because the platform's _RunEventLoop isn't running in the test. I guess I could call DeviceLayer::PlatformMgr().RunEventLoop(). But that does pull other things.

I've seen it done this way in other timer based test so i replicated that. Any other suggestion?

Reply:
I saw we use  ctx.DrainAndServiceIO() in other unit tests, could you check if we could re-use this or something similar? I think it will be quite painful if each unit test need to handle the event loop by itself.

Reply:
```
     * This DOES NOT ensure that all pending events are serviced to completion
     * (i.e timers, any ScheduleWork calls), but does:
     *
     * 1) Guarantee that every call will make some progress on ready-to-run
     *    things, by calling DriveIO at least once.
     * 2) Try to ensure that any ScheduleWork calls that happend directly as a
     *    result of message reception, and any messages those async tasks send,
     *    get handled before DrainAndServiceIO returns.
```
Looking at the implementation, I seem more targeted for the message layer. I  can give it a try, or use `DriveIO` but the end result is quite similar how what I do here.

Reply:
it now uses `GetIOContext().DriveIO()` instead of managing the event loop manually